File: src/core/security.ts
import type {
  CertificateDetails,
  ClientCertificateInfo,
  HttpAuthPrompt,
  SecurityPrompt,
  SecurityPromptResponse
} from '../shared/types'
import { newId } from '../shared/ids'
import type { Browser } from './browser'

export interface HttpCredentials {
  username: string
  password: string
}

/** What a host knows about an authentication challenge. */
export interface HttpAuthChallenge {
  host: string
  port: number
  realm: string
  /** Challenge scheme as the host reports it (any case); '' when it does not know. */
  scheme: string
  isProxy: boolean
  /** Whether the credentials travel over TLS. */
  secure: boolean
}

/**
 * Where remembered HTTP credentials live. The built-in store forgets everything when the browser
 * quits; the encrypted credential store can back this interface later.
 */
export interface HttpCredentialStore {
  get(key: string): HttpCredentials | null
  set(key: string, credentials: HttpCredentials): void
  delete(key: string): void
  clear(): void
}

export class SessionHttpCredentialStore implements HttpCredentialStore {
  private readonly entries = new Map<string, HttpCredentials>()

  get(key: string): HttpCredentials | null {
    return this.entries.get(key) ?? null
  }

  set(key: string, credentials: HttpCredentials): void {
    this.entries.set(key, credentials)
  }

  delete(key: string): void {
    this.entries.delete(key)
  }

  clear(): void {
    this.entries.clear()
  }
}

/** Credentials are remembered per protection space: proxy or server, host, port and realm. */
export function httpAuthKey(
  c: Pick<HttpAuthChallenge, 'host' | 'port' | 'realm' | 'isProxy'>
): string {
  return `${c.isProxy ? 'proxy' : 'server'}|${c.host.toLowerCase()}:${c.port}|${c.realm}`
}

/**
 * A challenge that arrives this soon after we answered one for the same protection space means
 * the answer was refused (the engine asks again for the same request).
 */
export const AUTH_RETRY_WINDOW_MS = 60_000

/** `host:port` of an https address (443 when it leaves the port out); null for any other URL. */
export function certificateSiteOf(url: string): string | null {
  try {
    const parsed = new URL(url)
    if (parsed.protocol !== 'https:' || !parsed.hostname) return null
    return `${parsed.hostname.toLowerCase()}:${parsed.port || '443'}`
  } catch {
    return null
  }
}

/** A certificate the user proceeded past, and the `ERR_CERT_*` code it failed with. */
export interface CertificateException {
  code: number
  certificate: CertificateDetails
}

/**
 * The certificates the user proceeded past this session, remembered as Chrome remembers them: per
 * site (host and port) and per certificate, so a site that changes its certificate asks again,
 * until the browser quits. Kept per container, so a private window's exceptions end with its
 * session. Nothing here is persisted. The hosts ask on every TLS handshake (Chromium keeps no
 * decision of its own in Electron), so this is the one place the answer lives.
 */
export class CertificateExceptions {
  private readonly allowed = new Map<string, CertificateException>()

  /**
   * Remember `certificate` for the site of `url`; false when `url` is not https or the
   * certificate has no fingerprint to remember it by.
   */
  allow(containerId: string, url: string, code: number, certificate: CertificateDetails): boolean {
    const key = this.key(containerId, url, certificate.fingerprint)
    if (!key) return false
    // Re-added last, so `exceptionFor` names the exception the latest load went ahead under.
    this.allowed.delete(key)
    this.allowed.set(key, { code, certificate })
    return true
  }

  isAllowed(containerId: string, url: string, fingerprint: string): boolean {
    const key = this.key(containerId, url, fingerprint)
    return key !== null && this.allowed.has(key)
  }

  /**
   * The exception a load of `url` in `containerId` goes ahead under, if the session has one for
   * its site: a committed https page of such a site was loaded over the excepted certificate (the
   * host asked, and this was the answer), unless the site has since put a valid one up. The
   * latest one when several certificates of the site were excepted.
   */
  exceptionFor(containerId: string, url: string): CertificateException | null {
    const site = certificateSiteOf(url)
    if (!site) return null
    const prefix = `${containerId}|${site}|`
    let found: CertificateException | null = null
    for (const [key, exception] of this.allowed) if (key.startsWith(prefix)) found = exception
    return found
  }

  /** The container's session ended (private browsing) or its site data was cleared: its exceptions go. */
  forgetContainer(containerId: string): void {
    for (const key of this.allowed.keys())
      if (key.startsWith(`${containerId}|`)) this.allowed.delete(key)
  }

  get size(): number {
    return this.allowed.size
  }

  private key(containerId: string, url: string, fingerprint: string): string | null {
    const site = certificateSiteOf(url)
    if (!site || !fingerprint) return null
    return `${containerId}|${site}|${fingerprint}`
  }
}

interface Pending {
  prompt: SecurityPrompt
  resolve: (response: SecurityPromptResponse | null) => void
}

/**
 * The prompts the engine needs an answer for before a request can go on: HTTP authentication
 * (Basic, Digest, NTLM, Negotiate, proxies) and client-certificate selection. Both are shown by
 * the chrome as a Zenium dialog; hosts only report the challenge and forward the answer.
 */
export class SecurityPromptService {
  private readonly pending: Pending[] = []
  /** Certificate chosen per host for this session; null means "continue without one". */
  private readonly certificateChoices = new Map<string, string | null>()
  /** When each protection space was last answered (to recognise refused credentials). */
  private readonly answeredAt = new Map<string, number>()
  /** The username last sent per protection space, offered again when it was refused. */
  private readonly lastUsername = new Map<string, string>()
  /**
   * One dialog per protection space: the engine challenges every request behind the same realm
   * (a page and its images), and they all wait for the one answer.
   */
  private readonly inFlight = new Map<string, Promise<HttpCredentials | null>>()
  private readonly certificateInFlight = new Map<string, Promise<number | null>>()
  /**
   * Server certificates the user proceeded past (the interstitial's "Proceed"). Separate from
   * `forgetSession`, which is about sign-ins; these end with the session or the browser.
   */
  readonly certificateExceptions = new CertificateExceptions()

  constructor(
    private readonly browser: Browser,
    readonly credentials: HttpCredentialStore = new SessionHttpCredentialStore(),
    private readonly now: () => number = Date.now
  ) {}

  list(): SecurityPrompt[] {
    return this.pending.map((p) => p.prompt)
  }

  /**
   * HTTP authentication. Resolves with the credentials to send, or null to give up (the page then
   * shows the server's 401). Remembered credentials are sent once without asking; when the same
   * space challenges again right away they were wrong, and the user sees the dialog.
   */
  async httpAuth(
    challenge: HttpAuthChallenge,
    tabId: string | null
  ): Promise<HttpCredentials | null> {
    const key = httpAuthKey(challenge)
    const waiting = this.inFlight.get(key)
    if (waiting) return waiting
    const now = this.now()
    const failedBefore = now - (this.answeredAt.get(key) ?? -Infinity) < AUTH_RETRY_WINDOW_MS
    const remembered = this.credentials.get(key)
    if (remembered && !failedBefore) {
      this.sent(key, remembered)
      return remembered
    }
    if (remembered) this.credentials.delete(key)
    const asking = this.askHttpAuth(key, challenge, tabId, failedBefore)
    this.inFlight.set(key, asking)
    try {
      return await asking
    } finally {
      this.inFlight.delete(key)
    }
  }

  private async askHttpAuth(
    key: string,
    challenge: HttpAuthChallenge,
    tabId: string | null,
    failedBefore: boolean
  ): Promise<HttpCredentials | null> {
    const prompt: HttpAuthPrompt = {
      id: newId('auth'),
      kind: 'http-auth',
      tabId,
      host: challenge.host,
      port: challenge.port,
      realm: challenge.realm,
      scheme: challenge.scheme.toLowerCase(),
      isProxy: challenge.isProxy,
      secure: challenge.secure,
      failedBefore,
      username: failedBefore ? (this.lastUsername.get(key) ?? '') : ''
    }
    const answer = await this.show(prompt)
    if (!answer || answer.kind !== 'http-auth') {
      this.answeredAt.delete(key)
      return null
    }
    const credentials = { username: answer.username, password: answer.password }
    if (answer.remember) this.credentials.set(key, credentials)
    this.sent(key, credentials)
    return credentials
  }

  private sent(key: string, credentials: HttpCredentials): void {
    this.answeredAt.set(key, this.now())
    this.lastUsername.set(key, credentials.username)
  }

  /**
   * Client certificate selection. Resolves with the index of the certificate to send, or null to
   * continue without one. Nothing is ever sent silently: the first request per host asks, and
   * the answer (including "none") holds for the rest of the session, as in Chrome.
   */
  async clientCertificate(
    host: string,
    certificates: ClientCertificateInfo[],
    tabId: string | null
  ): Promise<number | null> {
    const remembered = this.certificateChoices.get(host)
    if (remembered === null) return null
    if (remembered !== undefined) {
      const index = certificates.findIndex((c) => c.fingerprint === remembered)
      if (index >= 0) return index
    }
    if (!certificates.length) return null
    const waiting = this.certificateInFlight.get(host)
    if (waiting) {
      // Same host, same session: the answer to the first chooser applies here as well.
      await waiting
      return this.clientCertificate(host, certificates, tabId)
    }
    const asking = this.askClientCertificate(host, certificates, tabId)
    this.certificateInFlight.set(host, asking)
    try {
      return await asking
    } finally {
      this.certificateInFlight.delete(host)
    }
  }

  private async askClientCertificate(
    host: string,
    certificates: ClientCertificateInfo[],
    tabId: string | null
  ): Promise<number | null> {
    const answer = await this.show({
      id: newId('cert'),
      kind: 'client-certificate',
      tabId,
      host,
      certificates
    })
    const index =
      answer?.kind === 'client-certificate' && certificates[answer.index] ? answer.index : null
    this.certificateChoices.set(host, index === null ? null : certificates[index].fingerprint)
    return index
  }

  /** The chrome answered (or dismissed) a prompt. */
  respond(id: string, response: SecurityPromptResponse | null): void {
    const i = this.pending.findIndex((p) => p.prompt.id === id)
    if (i < 0) return
    const [entry] = this.pending.splice(i, 1)
    this.browser.state.commitVolatile()
    entry.resolve(response)
  }

  /** A tab went away: its prompts are moot. */
  cancelForTab(tabId: string): void {
    for (const p of this.pending.filter((p) => p.prompt.tabId === tabId))
      this.respond(p.prompt.id, null)
  }

  /** Forget the credentials and certificate choices of this session. */
  forgetSession(): void {
    this.credentials.clear()
    this.certificateChoices.clear()
    this.answeredAt.clear()
    this.lastUsername.clear()
  }

  /** Queue the prompt; the chrome shows it once its tab is the active one (tab-modal, like Chrome). */
  private show(prompt: SecurityPrompt): Promise<SecurityPromptResponse | null> {
    return new Promise((resolve) => {
      this.pending.push({ prompt, resolve })
      this.browser.state.commitVolatile()
    })
  }
}

File: android/app/src/main/kotlin/app/zen/chromium/Security.kt
package app.zen.chromium

import android.net.http.SslCertificate
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.security.KeyChain
import android.webkit.ClientCertRequest
import android.webkit.HttpAuthHandler
import android.webkit.WebViewDatabase
import org.json.JSONObject
import java.util.concurrent.Executors

/**
 * The security prompts the WebView delegates to the app. HTTP authentication (Basic, Digest,
 * NTLM) goes to the core's shared sign-in dialog and comes back through `auth.respond`; a
 * client-certificate request opens the system KeyChain chooser, and the alias picked is kept per
 * host and port until Zenium quits. A server certificate that fails verification is refused, and
 * the core's interstitial offers to proceed: the exception it records comes back through
 * `security.allowCertificate` ([certificateExceptions]) for the load that follows. Nothing is ever
 * sent without the user's say-so.
 */
class Security(private val host: PageHost) {
    private var seq = 0
    private val pendingAuth = HashMap<String, HttpAuthHandler>()
    private val certificateChoices = HashMap<String, String>()
    /** Server certificates proceeded past this session (the core's copy is the one that decides on the desktop). */
    val certificateExceptions = CertificateExceptions()
    private val main = Handler(Looper.getMainLooper())
    private val keychain = Executors.newSingleThreadExecutor { r -> Thread(r, "zenium-keychain") }

    fun onHttpAuth(tab: TabWebView, handler: HttpAuthHandler, hostName: String, realm: String) {
        val id = "auth_${++seq}"
        pendingAuth[id] = handler
        host.hostEvent(
            "auth.request",
            json("requestId" to id, "tabId" to tab.tabId, "host" to hostName, "realm" to realm, "url" to (tab.url ?: ""))
        )
    }

    /** The core's answer; no credentials means the user cancelled and the server's 401 page shows. */
    fun respondAuth(requestId: String, username: String?, password: String?) {
        val handler = pendingAuth.remove(requestId) ?: return
        if (username == null || password == null) handler.cancel() else handler.proceed(username, password)
    }

    fun onClientCertRequest(request: ClientCertRequest) {
        val key = "${request.host}:${request.port}"
        val remembered = certificateChoices[key]
        if (remembered != null) {
            proceed(key, remembered, request)
            return
        }
        KeyChain.choosePrivateKeyAlias(
            host.activity,
            { alias ->
                main.post {
                    if (alias == null) {
                        request.cancel()
                    } else {
                        certificateChoices[key] = alias
                        proceed(key, alias, request)
                    }
                }
            },
            request.keyTypes, request.principals, request.host, request.port, null
        )
    }

    /** KeyChain lookups block on another process: off the main thread, then back to answer. */
    private fun proceed(key: String, alias: String, request: ClientCertRequest) {
        keychain.execute {
            val privateKey = runCatching { KeyChain.getPrivateKey(host.activity, alias) }.getOrNull()
            val chain = runCatching { KeyChain.getCertificateChain(host.activity, alias) }.getOrNull()
            main.post {
                if (privateKey != null && chain != null) {
                    request.proceed(privateKey, chain)
                } else {
                    certificateChoices.remove(key)
                    request.cancel()
                }
            }
        }
    }

    /** Settings → Security → Forget now: the WebView's own copies go too. */
    fun forgetSession() {
        certificateChoices.clear()
        runCatching { WebViewDatabase.getInstance(host.activity).clearHttpAuthUsernamePassword() }
    }

    /** The user proceeded past the interstitial for the site of `url` over this certificate (the core recorded it). */
    fun allowCertificate(containerId: String, url: String, fingerprint: String) {
        certificateExceptions.allow(containerId, url, fingerprint)
    }

    /** Whether a request of a `containerId` page may go ahead over the certificate `fingerprint` names. */
    fun certificateAllowed(containerId: String, url: String, fingerprint: String?): Boolean =
        certificateExceptions.isAllowed(containerId, url, fingerprint)

    /**
     * The container's data went (deleted, private session over, cookies cleared): its exceptions
     * go, and the WebView's own memory of every `proceed()` on the profile with them, so the next
     * load asks again. `clearSslPreferences` is a view's call but clears the profile's store; a
     * container without a live page has its profile deleted with the data (`Profiles.clear`).
     */
    fun forgetCertificates(containerId: String) {
        certificateExceptions.forgetContainer(containerId)
        host.tabs.all().firstOrNull { it.containerId == containerId }?.clearSslPreferences()
    }

    fun shutdown() {
        keychain.shutdownNow()
    }

    companion object {
        /**
         * What the interstitial shows of a refused certificate (`CertificateDetails` in
         * `src/shared/types.ts`), with the fingerprint the exception is keyed by; null without a
         * certificate. Names are the common name, or the organisation when there is none.
         */
        fun describeCertificate(certificate: SslCertificate?): JSONObject? {
            if (certificate == null) return null
            val der = derOf(certificate)
            return json(
                "subjectName" to nameOf(certificate.issuedTo),
                "issuerName" to nameOf(certificate.issuedBy),
                "validStart" to (certificate.validNotBeforeDate?.time ?: 0L),
                "validExpiry" to (certificate.validNotAfterDate?.time ?: 0L),
                "fingerprint" to (der?.let(CertificateExceptions::fingerprintOf) ?: "")
            )
        }

        /** The certificate's fingerprint (`CertificateExceptions.fingerprintOf`), or null when its bytes are out of reach. */
        fun fingerprintOf(certificate: SslCertificate?): String? =
            certificate?.let(::derOf)?.let(CertificateExceptions::fingerprintOf)

        /** The DER bytes of a WebView certificate: its X509 from API 29, the saved state's copy before. */
        private fun derOf(certificate: SslCertificate): ByteArray? = runCatching {
            val x509 = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) certificate.x509Certificate else null
            x509?.encoded ?: SslCertificate.saveState(certificate).getByteArray("x509-certificate")
        }.getOrNull()

        private fun nameOf(name: SslCertificate.DName?): String =
            name?.cName?.takeIf { it.isNotEmpty() } ?: name?.oName?.takeIf { it.isNotEmpty() } ?: name?.dName ?: ""
    }
}
